### DISCUSSION TITLE
How can I dynamically register the menu?

### DISCUSSION THREAD

**Discussion by baoqing1124** [eclipse-theia/theia#13716] [Q&A]
I need to register the menu dynamically during the use of the application, but I have tested that the menu cannot be displayed when I register the menu asynchronously. What should I do to achieve this function? Can anyone help me?
![31a22891126ae9a389a3ef778825fb1](https://github.com/eclipse-theia/theia/assets/15626996/228d0499-a0ea-4872-91c2-1255c0abc2cc)
![e0266e0991223cb9a33921435942ce3](https://github.com/eclipse-theia/theia/assets/15626996/8c461999-418b-4e5e-bcc9-51c0baa77894)

**Comment by msujew** [comment]
Hey @baoqing1124,

Can you provide a reproducible example? We're performing dynamic registration of menu actions (see [here for an example](https://github.com/eclipse-theia/theia/blob/32f393a3385927ed79708b4e1a2faf78806f1fd3/packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts#L347)) all the time in the framework. It's supposed to work.

**Reply by baoqing1124** [reply]
hi~ @msujew 
This is an [example](https://github.com/baoqing1124/theia-demo) based on theia master, I modified the the code in "examples\api-samples\src\browser\menu\sample-menu-contribution.ts" at line 277.

**Reply by msujew** [reply]
Looking into the code, it actually looks like the menu is filled/created once when it is made visible. Every other modification to the menu is ignored afterwards (unless you recreate it by changing the `window.menuBarVisibility` setting). I've created https://github.com/eclipse-theia/theia/issues/13719 to track this.

**Comment by kittaakos** [comment]
> What should I do to achieve this function? Can anyone help me?

@baoqing1124, for now, you can see how the Arduino IDE solves this: https://github.com/arduino/arduino-ide/blob/aa9b10d68e6653b7f40ab920173c7964aaefbfa4/arduino-ide-extension/src/common/main-menu-manager.ts#L7

If you want to unregister submenus, there is a non-API way to do it: https://github.com/arduino/arduino-ide/blob/aa9b10d68e6653b7f40ab920173c7964aaefbfa4/arduino-ide-extension/src/browser/menu/arduino-menus.ts#L198
